fix(frontend): create two tests for the unauthorised component

Slav666 opened the pull request.

Commit:
fix(frontend): create two tests for the unauthorised component

marksmall commented on client/src/app/unauthorized.component.test.tsx:
```diff
@@ -0,0 +1,24 @@
+  it('should render unauthorized component', () => {
+    render(<Unauthorized />);
+
+    expect(screen.getByRole('heading', { name: /Unauthorized/i })).toBeInTheDocument();
```
by using the `//i` syntax, which means regex case agnostic, then you shouldn't capitalize anything, keep it all lowercase. That is how Kent Dodds and other bloggers do it.

marksmall commented on client/src/app/unauthorized.component.test.tsx:
```diff
@@ -0,0 +1,24 @@
+
+    userEvent.click(screen.getByRole('button', { name: 'Go Back' }));
+
+    await waitFor(() => expect(goBack).toBeCalled());
```
I think `toBeCalled` is an alias, we have never used this in any of our code, always use `toHaveBeenCalled`. If you do find other code doing this, then flag it for change, as it is wrong. I have no added an eslint rule that catches this, so you will see an error in future.

marksmall commented on client/src/app/unauthorized.component.tsx:
```diff
@@ -1,12 +1,10 @@
+  goBack?: () => void;
+}
 
+export const Unauthorized: FC<Props> = ({ goBack }): ReactElement => {
```
You can now make this an `implicit` return, rather than `explicit`. You should only use explicit returns, when you have functions/hooks etc, like you did in the previous code. This is another pattern I want to enforce, so  our code is as similar as possible, making everyone's lives easier.

marksmall reviewed (changes requested):
Couple of comments, but a far better PR than the previous one.

Commit:
fix(frontend): add requested changes to the code and update a git ignore file

marksmall commented:
Instead of making commits of one line saying `add requested changes....` which is really cryptic. If I go back to the commits in 3 weeks time, how do I find the PR that the commit relates to, where there is the conversation? Every request for change should be their own `fix` commit. The `title` should be a short version of what was asked to be changed and the body, hopefully should be able to get that from the PR conversation and then what you did and why. Single line commits are not acceptable, I will just delete code that does this. If you wrote the code, you must be able to explain it in plain english. I know English is a bad language, maybe write it in Polish and translate it to English, Google Translate is pretty good. Basically, it's up to you to explain what you did and why, not rely on others figuring it out.

marksmall commented on .vscode/settig.json:
The name of this file is wrong and will not work, simple typo, it should be `settings.json`. I've not gon through this file, but I hope you got it from my `orbison` code, as there will have been changes. If you have just used an old one from say the blogs app, it could easily be missing stuff.

marksmall commented on client/src/app/unauthorized.component.test.tsx:
```diff
@@ -0,0 +1,24 @@
+    const goBack = vi.fn();
+    render(<Unauthorized goBack={goBack} />);
+
+    userEvent.click(screen.getByRole('button', { name: 'Go Back' }));
```
I thought I had previously linked to how to use the userEvent, please use it how the docs use it, not in some random stackoverflow, I got caught out by doing that. See https://testing-library.com/docs/user-event/intro/, it has an example. So there is setting up the user, then using the user to click things, it should be pretty self-explanatory, but if not, just ask.

One of the things is to await the click, that should probably negate the expect also having to be await'd, making it more linear, async stuff is hard.

marksmall reviewed (changes requested):
Sorry, couple of more things, but what your producing is good, just a tad outdated, by like 6 months.... Software moves fast.

Commit:
fix(frontend): add better commit message, refactor userEvent code and change content and name of setting.json file

The purpose of the first test is to verify that the component is rendered correctly by examining the contents of the component. I want to achieve this by checking the correct content of the heading, paragraph and button tags.The purpose of the second test is to check whether the goBack function will be called after clicking on the go-back button. I changed the code in this test to invoke userEvent.setup() before rendering the component which is recommended way according to the https://testing-library.com/docs/user-event/intro/ documentation. I renamed the file from setting.json to settings.json, it was just a spelling error. I changed the content of this file, which is now identical to the Orbison application.

RoryMacGregor88 commented on client/src/app/unauthorized.component.test.tsx:
```diff
@@ -0,0 +1,25 @@
+  });
+
+  it('should move one step backward from unauthorized page when clicks on the "Go Back" button', async () => {
+    const user = userEvent.setup();
```
I'm not familiar with this use of `userEvent`, what is this doing?

Slav666 replied:
There is a link to the documentation which explains that: https://testing-library.com/docs/user-event/intro/ Mark  just asked me to do it in that way.

RoryMacGregor88 replied:
Ok cool, good to know

marksmall replied:
This is not something I am asking anyone to do, this is what the docs say, don't confuse the two. How to use `userEvent` has changed fairly recently, so everyone has to update their knowledge on how to use it.

RoryMacGregor88 commented on client/src/app/unauthorized.component.test.tsx:
```diff
@@ -0,0 +1,25 @@
+    const goBack = vi.fn();
+    render(<Unauthorized goBack={goBack} />);
+
+    await user.click(screen.getByRole('button', { name: /go back/i }));
```
I've stopped using regex to match things in tests, because it's not testing for text that should be capitalised. Like if it's supposed to be `Go Back` but I type `gO bACk` it would still pass.

stevefaeembra replied:
the /i makes this case insensitive, good practice is to lowercase search term and add this

RoryMacGregor88 replied:
For things like search terms or forms, where we'd be lowercasing everything anyway I imagine it would be redundant to make matchers case-insensitive.

But for things like buttons, typos could slip through. Like if I typed `go Back` as the button text, this test wouldn't catch the mistake because of the case-insensitive regex.

marksmall replied:
I have yet to see an example by the creators of testing-library use plain strings. I don't think unit/integration testing should care about typos like this. That would more likely be the place to test using **snapshots** or E2E testing, or even **stories**. Use the right kind of test to test what is appropriate. That being said, I originally thought the same as Rory, but I seem to be in the minority, if you look at blogs|tutorials|the docs etc.

marksmall replied:
What matters most is we are consistent. I don't want some using plain strings and others using regex, do what the docs suggets, don't make it up as you go along @RoryMacGregor88. If you disagree, fine, state reasons, argue over it and agree, but don't arbitrarily, do things different just cause you disagree.

RoryMacGregor88 commented on client/src/app/unauthorized.component.tsx:
```diff
@@ -1,25 +1,21 @@
 
-import { useNavigate } from 'react-router-dom';
+interface Props {
+  goBack?: () => void;
```
This shouldn't be optional, as it is always being passed.

Slav666 replied:
This is optional because the first test doesn't need the goBack props. Do I need to refactor the first test with goBack props?

RoryMacGregor88 replied:
Yeah I'd pass it in every test. I used to only pass the props needed for each test, but I realised that having missing props in tests could lead to unexpected behaviour, so now I add all props in all tests.

marksmall replied:
The test shouldn't drive the component, only things that are optional during actual live running should be optional. 

Remember when developing, interfaces|types tell the developer what they need to pass during **actual usage**, not during testing, where you may only be focussing on a part of the functionality. So, while the one single test might not need it to work, due to your focus being only on what DOM elements are present. The component would be broken, as when the user clicked on it, it would do nothing. It is also possible, as Rory states, that not providing the prop, could have some consequences you didn't foresee.

Commit:
fix(frontend): add some cosmetic changes to the code

I changed the name of the settings.json file, it was a spelling mistake. I added missed the goBack props to the first test. I removed regex to match things in tests.

Commit:
renamed setting.json file

Commit:
fix(frontend): replace plain strings in second test

I refactored the second test code in places where I use plain strings instead of using regex.

Commit:
fix(frontend): replace plain strings in second test

The title of previous commit was wrong. I refactored some code in first test, not the second one. I refactor secend test code in places where I use plain strings instead of using regex.
